### ISSUE TITLE
Intermittent FileNotFoundError in np.save on mounted drive

### ISSUE THREAD

**Issue by leopd** [numpy/numpy#14626]
Sometimes (not always) when calling `np.save` on an object (using pickle), I get a FileNotFoundError.  I see it when the output is being written to a USB-mounted drive formatted with HFS+.  I don't see it on a local drive formatted with ext4, nor on an SMB-mounted remote drive.  I haven't exhaustively checked different filesystem types or combinations to further narrow down the dependency on the exact system configuration, but there's clearly something wrong.  Ubuntu 18.04 (full system info below).  This has been bothering me for a while, and I finally narrowed it down to a simple repro case to post here.

Just because somebody will ask, no I'm not unplugging the USB drive in the middle of this.

### Reproducing code example:

```python
import numpy as np
  
OUT_FILE = "/vol/hfsplus-drive/test-bug/out.npy"

print(f"Generating data...")
d = {}
for i in range(20000):
    d[str(i)] = np.asarray([0])
for attempt in range(100):  # Try this 100 times because sometimes it works
    print(f"Attempt {attempt} writing...")
    np.save(OUT_FILE, d, allow_pickle=True)
    print("Done")
```

### Error message:

```
Generating data...
Attempt 0 writing...
Done
Attempt 1 writing...
Traceback (most recent call last):
  File "/home/leo/anaconda3/lib/python3.7/site-packages/numpy/lib/npyio.py", line 536, in save
    pickle_kwargs=pickle_kwargs)
  File "/home/leo/anaconda3/lib/python3.7/site-packages/numpy/lib/format.py", line 629, in write_array
    pickle.dump(array, fp, protocol=2, **pickle_kwargs)
FileNotFoundError: [Errno 2] No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bug.py", line 11, in <module>
    np.save(OUT_FILE, d, allow_pickle=True)
  File "/home/leo/anaconda3/lib/python3.7/site-packages/numpy/lib/npyio.py", line 539, in save
    fid.close()
FileNotFoundError: [Errno 2] No such file or directory
```

### Numpy/Python version information:

```
>>> import sys, numpy; print(numpy.__version__, sys.version)
1.16.3 3.7.1 (default, Dec 14 2018, 19:28:38) 
[GCC 7.3.0]
```

Also, since it seems likely relevant: Ubuntu 18.04:

```
% uname -a
Linux ubuntu-laptop 5.0.0-29-generic #31~18.04.1-Ubuntu SMP Thu Sep 12 18:29:21 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by seberg** [comment]
I do not think this has anything to do with numpy, or even python. The `fp` is the opened file, so opening the file was actually successful. I suspect you could reproduce it on other harddrives if you specifically unmount them. Maybe check your `dmesg` for unmounting/usb error events?

**Comment by leopd** [comment]
Some supporting but not conclusive evidence that this is a numpy issue, not a python/pickle issue, is that if I try to do something similar without numpy, it works reliably:

```python
import pickle
  
OUT_FILE = "/vol/hfsplus-drive/test-bug/out.pickle"

print(f"Generating data...")
d = {}
for i in range(20000):
    d[str(i)] = [0]
for attempt in range(100):  # 100 times this never fails
    print(f"Attempt {attempt} writing...")
    with open(OUT_FILE, "wb") as f:
        pickle.dump(d, f, protocol=2)
    print("Done")
```

**Comment by leopd** [comment]
@seberg Nothing in `dmesg`.

Yes, if you unmount the drive in the middle of writing, bad things will happen.  Again, I'm not unmounting the drive.  Neither is anything else.

**Comment by leopd** [comment]
The fact that it works intermittently implies to me it's a race condition or some other timing interaction.  A bit more digging and I've isolated it to the USB system, and not a function of the filesystem type.

WORKS:
* ext4 filesystem on NVMe disk
* Loopback filesystem using HFS+ on the local fast NVMe disk
* SMB-mounted remote filesystem

FAILS:
* HFS+ filesystem on magnetic USB disk (original post)
* Loopback ext4 filesystem on the magnetic USB HFS+ disk.
* Vfat filesystem on a USB solid-state disk.
* ext4 filesystem on a USB solid-state disk.
* HFS+ filesystem on a USB solid-state disk.

**Comment by leopd** [comment]
Closing this because I'm convinced the root cause is a hardware issue on my machine, as I'm seeing some other inappropriate behavior on these drives.  Sorry for the noise.  Sure wish Ubuntu would log some error messages when the hardware is having problems, but I really can't find anything.

One more oddity I've observed, in case anybody else sees this or something similar: it only seems to happen if the path includes a symlink.  Go figure.
